File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.cpp
#include <bits/stdc++.h>

using namespace std;

vector<int> solution(vector<string> operations) {
    
    multiset<int> pq;
    
    for (string& operation : operations) {
        
        string cmd;
        int num;
        
        stringstream ss(operation);
        ss >> cmd;
        ss >> num;
        
        // 숫자 삽입
        if (cmd == "I") {
            pq.insert(num);
        }
        // 큐가 비었을 때
        else if (pq.empty()) {
            continue;
        }
        // 최대값 제거
        else if (num == 1) {
            pq.erase(*pq.rbegin());
        }
        // 최소값 제거
        else if (num == -1) {
            pq.erase(*pq.begin());
        }
        
    }
    
    if (pq.empty()) {
        return {0, 0};
    }
    return {*pq.rbegin(), *pq.begin()};
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;

class Solution {
    public int[] solution(String[] operations) {
        
        TreeMap<Integer, Integer> pq = new TreeMap<>();
        
        for (String operation : operations) {
            
            String cmd = operation.split(" ")[0];
            int num = Integer.parseInt(operation.split(" ")[1]);
            
            // 숫자 삽입
            if (cmd.equals("I")) {
                pq.put(num, pq.getOrDefault(num, 1));
            }
            // 큐가 비어있는 경우
            else if (pq.isEmpty()) {
                continue;
            }
            // 최대값 제거
            else if (num == 1) {
                int max = pq.lastKey();
                if (pq.get(max) == 1) {
                    pq.remove(max);   
                } else {
                    pq.put(max, pq.get(max) - 1);
                }
            } 
            // 최소값 제거
            else if (num == -1) {
                int min = pq.firstKey();
                if (pq.get(min) == 1) {
                    pq.remove(min);   
                } else {
                    pq.put(min, pq.get(min) - 1);
                }
            }
        }
        
        if (pq.isEmpty()) {
            return new int[] {0, 0};
        }
        
        return new int[] {pq.lastKey(), pq.firstKey()};
    }
}